services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java

```diff
@@ -0,0 +1,254 @@
+import com.futurewei.alcor.common.exception.NetworkTypeInvalidException;
+import com.futurewei.alcor.common.exception.VlanRangeNotFoundException;
+import com.futurewei.alcor.vpcmanager.dao.GreRepository;
+import com.futurewei.alcor.vpcmanager.dao.VlanRangeRepository;
+import com.futurewei.alcor.vpcmanager.dao.VlanRepository;
+import com.futurewei.alcor.vpcmanager.dao.VxlanRepository;
+import com.futurewei.alcor.vpcmanager.service.SegmentService;
+import com.futurewei.alcor.web.allocator.VlanKeyAllocator;
+import com.futurewei.alcor.web.entity.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+@Service
+public class SegmentServiceImpl implements SegmentService {
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @Autowired
+    private VlanRangeRepository vlanRangeRepository;
+
+    @Autowired
+    private VlanRepository vlanRepository;
+
+    @Autowired
+    private VxlanRepository vxlanRepository;
+
+    @Autowired
+    private GreRepository greRepository;
+
+    @Override
+    public Long addVlanEntity(String segmentId, String vlanId, String networkType) throws DatabasePersistenceException, CacheException {
+        try {
+            NetworkVlanType vlan = new NetworkVlanType();
+            // check if range exist in db
+            Map<String, NetworkVlanRange> map = this.vlanRangeRepository.findAllItems();
+            if (map.size() == 0) {
+                String rangeId = UUID.randomUUID().toString();
+                NetworkVlanRangeRequest request = new NetworkVlanRangeRequest(rangeId, segmentId, networkType, NetworkType.VLAN_FIRST_KEY, NetworkType.VLAN_LAST_KEY);
+                this.vlanRangeRepository.createRange(request);
+            }
+
+            String rangeId = "";
+            for (Map.Entry<String, NetworkVlanRange> entry : map.entrySet()) {
+                rangeId = entry.getKey();
+            }
+
+            Long key = this.vlanRangeRepository.allocateVlanKey(rangeId);
+
+            //vlan.setMtu(mtu);
+            vlan.setSegmentId(segmentId);
+            vlan.setVlanId(vlanId);
+            vlan.setKey(key);
+
+            this.vlanRepository.addItem(vlan);
+            logger.info("Allocate vlan key success, key: " + key);
+            return key;
+        } catch (Exception e) {
+            this.vlanRepository.deleteItem(vlanId);
+            logger.info("Allocate vlan key failed");
+            throw new DatabasePersistenceException(e.getMessage());
+        }
+
+    }
+
+    @Override
+    public Long addVxlanEntity(String segmentId) throws DatabasePersistenceException {
+        try {
+            String vxlanId = UUID.randomUUID().toString();
+            NetworkVxlanType vxlan = new NetworkVxlanType();
+            //vxlan.setMtu(mtu);
```

Comment:
Please uncomment this. We do need MTU.

Reply:
Yes, but where can we get mtu? segmentRequestObject doesn't include it